### PR TITLE
[WIP] Convert to a minor mode (Fix #18)

### LINKED ISSUE #18: (require 'outshine) is effectful
Browsing through outshine.el, I noticed that just loading the package is effectful[1] --- there are `(put)`s and `(defadvice)`s at toplevel, among other things.  IMHO, it would be better, at least for debugging, if loading had no effects beyond defining new symbols.  

[1] I'd agree that "effectful" in elisp is hard to define, but here it should be understood as doing anything that isn't `def*`  (defun, defvar, defcustom, defmacro...)

### PR DESCRIPTION
**Pull request by thblt**
This is a WIP PR to convert the activation mechanism from a hacky override to a regular minor mode.  

Calvin behaving

This PR is to track advancement, especially in case I give up before it's done (which may happen...), so anyone who feels like finishing the work could hopefully have half the job already done and the remaining part in a neat TODO  list.

# TODO/FIXME

## Namespace isolation issues 

These issues concern isolation = not touching anything not defined in `outshine.el` and not in the `outshine-` pseudo namespaces (these two definitions should be strictly equivalent)

- [x] `outshine-set-local-outline-regexp-and-level` sets `outline-level`[1]
- [x] `outshine-set-local-outline-regexp-and-level` sets `outline-regexp`[1]
- [x] `outshine-set-local-outline-regexp-and-level` sets `outline-heading-end-regexp`[1]

[1] I don't think this can be avoided, but commit "Save and restore outline- variables we manipulate"  introduces a mechanism to revert theses variables to their original values when deactivating outshine.  

- [ ] **IDEA** Would it be useful to save the modified values of these variables and to restore them only if they weren't modified externally?  (I don't think so, but maybe)

## Purity issues

These issues concern purity:

 1. `(require 'outshine)` should have no effect except defining *new* symbols in the `outshine-` pseudo namespace.
 2. `(outshine-minor-mode 1)` should have no effect outside the current buffer.
 3. `(progn (outshine-minor-mode 1) (outshine-minor-mode 0))` should have absolutely no effects.

- [x] The first invocation of `(outshine-minor-mode 1`) adds `outshine--outline-minor-mode-hook` on `outline-minor-mode-hook`.  This is hard to avoid if we want sane behavior, and I don't consider it a bug.  (The purpose of this function is to deactivate outshine when outline gets deactivated)
- [X] **SEVERE** What is the `defadvice org-store-log-note` for?  It isn't local, but it shouldn't be activated by `(require 'outshine)`.[2]  
- [x] What is the series of `(put SYMBOL :advertised-binding BINDING)` useful for?
- [ ] **SEVERE** https://github.com/alphapapa/outshine/pull/19#issuecomment-428995884

[2] I still have no idea what that's for, but I wrapped the advice in `(when outshine-minor-mode` and moved the `defadvice` to `outshine--minor-mode-activate`

## Compatibility

 - [ ] Fix Outorg compatibility.  Outorg calls `outshine-hook-function`.

## Miscellaneous

- [X] Should Outshine work with `outline-mode` as well (the major mode)?  In my tests, it doesn't work.   (No: see https://github.com/alphapapa/outshine/pull/19#issuecomment-402059191)
- [ ] Slightly unrelated: outshine does not really *depend* on Outorg, but it knows how to use it.  Could we remove the dependency and replace it  with `(require 'outorg t) ... (when (featurep 'outorg) ...)`?

### CODE REVIEW

**Comment by casouri** [comment]
I dont't think outshine needs to work with outline-mode. I think it only needs to work with outline-minor-mode since it serves primarily all the programming major modes.

**Comment by thblt** [comment]
Thanks for the clarification, I marked the point as solved (and opened two new ones :)

**Comment by thblt** [comment]
Also, but just so Github creates explicit links between related things: this is an attempt at solving #18.

**Comment by casouri** [comment]
I have to express my admiration of you for refactoring such a large package...

**Comment by alphapapa** [comment]
> I have to express my admiration of you for refactoring such a large package...

Indeed. While having no expectation that Thibault will necessarily complete the refactoring, or do so anytime soon, I am very grateful for his undertaking to start this work. It is not a glorious or very fun task, but it is much needed and will be much appreciated!

File: outshine.el [hunk 1]
```diff
@@ -1614,9 +1629,23 @@ function was called upon."
...
 
-;; ;; add this to your .emacs
-;; (add-hook 'outline-minor-mode-hook 'outshine-hook-function)
+(defun outshine--minor-mode-deactivate ()
+  "This function is private, use `outshine-minor-mode' to toggle Outshine.")
+
+;;;###autoload
+(defun outshine-hook-function ()
+  "DEPRECATED, use `outshine-minor-mode'."
+  (error "`outshine-hook-function' has been deprecated, use `outshine-minor-mode'"))
```
Comment (alphapapa) [inline outshine.el]: Would it be good to have an interim release where this still worked but gave a message with sleep-for advising of its becoming obsolete?

This all looks fantastic.  Thank you so much.

File: outshine.el [hunk 2]
```diff
@@ -1570,6 +1553,25 @@ Don't use this function, the public interface is
   ;; Install deactivation hook
   (add-hook 'outline-minor-mode-hook 'outshine--outline-minor-mode-hook)
 
+  ;; Advise org-store-log-note
+  (defadvice org-store-log-note (around org-store-log-note-around activate)
```
Comment (alphapapa) [inline outshine.el]: It would probably be good to convert this to use advice-add and advice-remove, if you are interested. :)

File: outshine.el [hunk 3]
```diff
@@ -1550,6 +1550,9 @@ world outside of the Org major-mode."
 Don't use this function, the public interface is
 `outshine-minor-mode'."
 
+  ;; Ensure outline is on
+  (outline-minor-mode 1)
```
Comment (alphapapa) [inline outshine.el]: Great idea!

File: outshine.el [hunk 4]
```diff
@@ -1608,7 +1622,12 @@ Don't use this function, the public interface is
       (outshine-fontify-headlines out-regexp))))
 
 (defun outshine--minor-mode-deactivate ()
-  "This function is private, use `outshine-minor-mode' to toggle Outshine.")
+  "Deactivate Outshine.
+
+Don't use this function, the public interface is
+`outshine-minor-mode'."
+  ;; Restore variables
+  (cl-mapc 'set outshine-protected-variables outshine-protected-variables-values))
```
Comment (alphapapa) [inline outshine.el]: A clever solution to a great idea!

File: outshine.el [hunk 5]
```diff
@@ -452,6 +452,9 @@ A comment subtree does not open during visibility cycling.")
 (defvar-local outshine-protected-variables-values nil
 "The values of variables defined by `outshine-protected-variables'.")
```
Comment (alphapapa) [inline outshine.el]: Great!

File: outshine.el [hunk 6]
```diff
@@ -1646,7 +1646,8 @@ Don't use this function, the public interface is
 ;;;###autoload
 (defun outshine-hook-function ()
   "DEPRECATED, use `outshine-minor-mode'."
-  (error "`outshine-hook-function' has been deprecated, use `outshine-minor-mode'"))
+  (warn "`outshine-hook-function' has been deprecated, use `outshine-minor-mode'")
```
Comment (alphapapa) [inline outshine.el]: Excellent, you already thought of this! :)

File: outshine.el [hunk 7]
```diff
@@ -1573,7 +1573,8 @@ Don't use this function, the public interface is
 `outshine-minor-mode'."
 
   ;; Ensure outline is on
-  (outline-minor-mode 1)
+  (unless outline-minor-mode
```
Comment (alphapapa) [inline outshine.el]: What if you only enable the mode if it's not already active?

Comment (thblt) [reply outshine.el]: I've just done this in c8a2c7e, and I've been testing it for a few days, it seems to work (at last!)

File: outshine.el (hunk 1 above)
Comment (thblt) [reply outshine.el]: I've replaced `(error` with `(warn` in a previously pushed commit, because indeed there's no reason to break compatibility there.

**Comment by thblt** [comment]
@casouri @alphapapa  Thanks a lot for the kind words, much appreciated :)

@alphapapa I'd be happy to assist you in maintaining outshine if you need a hand, of course!

**Comment by JohnLunzer** [comment]
Is there anything holding this PR back from being merged with master? Just thought I'd poke at it.

**Comment by alphapapa** [comment]
Well, it's a major change, and there are still some unchecked boxes on Thibault's to-do lists up there.  I expect that he will say when he thinks it's ready to merge.  Feel free to test the branch and let us know how it works for you.

**Comment by abradd** [comment]
It seems that when outshine is loaded, but the minor mode isn't running in an org-note buffer, the capture of the note doesn't complete. Pressing C-c C-c or C-c C-k has no effect. I noticed that there is advice for `org-store-log-note' even when the outshine-minor-mode hasn't been invoked in the note buffer. I don't know much about advising functions so I have no idea if this is expected behaviour.

When debugging it seems that calling C-c C-c in the note buffer doesn't make it past the advising function to `org-store-log-note' when outshine-minor-mode isn't enabled. If i enable outshine-minor-mode in the note buffer, C-c C-c correctly parses the note.

I'm running outshine revision c8a2c7e, org mode from master 2fda33bfe in emacs 27.0.50.

**Comment by alphapapa** [comment]
@abradd Thanks for reporting that.  We should only advise that function when the minor mode is enabled, and preferably only locally, within the buffers it's enabled in.

**Comment by tj64** [comment]
Hey guys,
congratulations for this refactoring effort, very good job with a valuable goal!
And I'm really positively surprised how many people react to your initiative (as outshine users).

Just for the record, here a few historical remarks:

1. outshine started out as a merge of existing outline extensions (but extending outline-minor-mode instead of the major mode). Any "redefinition" of outline functionality from these extensions is only "by accident", because when these extensions were written the functions did not exist in core outline mode. Very good move to get rid of them now.

2. outshine started out as extension to outline-minor-mode for the historical reasons described above.  
Since it has grown over time to quite a big package, I like the idea very much to convert it into an independent well-behaved minor-mode.

3. Its true that outshine does not depend on outorg, but uses it to bring more Org functionality to outshine without reimplementing it.

4. 1+ from me for replacing "outshine-outline- ..." prefix with just "outshine-..."

Great work, looking forward to see it progress.
I'll be travelling for some time now, but any questions about "hard to comprehend" parts of outshine are welcome, hope I remember whats it was about.

Cheers
Thorsten

**Comment by yuhan0** [comment]
Any update on the progress of this PR? I've been using this fork for the past few months and made many local edits and improvements to it - would be nice to contribute some of that back here without conflicts :) 
May I suggest that this PR be merged as-is into a new "develop" branch that would make it easier to submit further PRs here instead of against @thblt's fork? 
I believe this also breaks compatibility with other packages such as helm-navi which depend on outshine, so having it as a separate branch would make things easier for development.

**Comment by thblt** [comment]
@yuhan0 I have another suggestion. I'm afraid if we don't release one day or another, this PR is going to rot (by my fault!) I've reviewed the changes and the conversation, and except for  https://github.com/alphapapa/outshine/pull/19#issuecomment-428995884, which, I think, is solved by 
https://github.com/alphapapa/outshine/pull/19/commits/5d745bd944623be71de919b1463fd92f1468a4f2, everything seems good. So why wouldn't we just merge this on master? (tbh I'm also suggesting this because I'm on holidays for two weeks, so I'll have enough free time to address issues)

(There's still one unchecked box: the circular dependency on outorg, but this is really not critical and not even actually related to this PR, so can be solved separately)

@alphapapa , what do you think?

**Comment by alphapapa** [comment]
Hi Thibault,

Good to hear from you again.  :)  If you think it's ready to merge, then I guess it probably is!  However...

1.  Outorg circular dependency.  Will merging this cause Outorg users to start getting errors on load?  If so, I think we have to fix that before merging this.

2.  I think this major of a change requires a full major version number bump, so it should be Outshine 3.0.

3.  If Outorg requires changes to work with the new Outshine, it should have its major version number bumped to match.

4.  Same for navi-mode.

5.  After doing all of those and pushing to `master` for each that needs it, we should make an announcement in the appropriate places.

What do you think about those issues?

Thanks for your work on this!

**Comment by thblt** [comment]
Hi Adam,

> Outorg circular dependency. Will merging this cause Outorg users to start getting errors on load? If so, I think we have to fix that before merging this.

I haven't changed anything regarding this.  The issue is that Outshine and Outorg both pretend they depend on the other, and I don't think this is the case.  But it's outside the scope of this PR.

> I think this major of a change requires a full major version number bump, so it should be Outshine 3.0.

Agreed, will do this when I rebase on your master.

> If Outorg requires changes to work with the new Outshine, it should have its major version number bumped to match.
>
> Same for navi-mode.

There shouldn't be a need, since `outshine-hook-function` still exists, it just generates a deprecation warning.  This is not perfect, but imo it's better to get outshine 3 out first then update the packages depending on it (better to call a deprecated function than a non-existing one, which will happen if for some reason someone updates outorg or navi before outshine) 

> After doing all of those and pushing to master for each that needs it, we should make an announcement in the appropriate places.

Here's a draft:

# [Announce] Outshine 3.0 released

The [Outshine](https://github.com/alphapapa/outshine) packages brings Org-Mode headings and structural editing commands to non-Org Emacs buffers, especially source code buffers.

We just released Outshine 3.0, the new major version of Outshine.  This version shouldn't change anything from the user's point of view: instead, it focuses on updating Outshine implementation to modern Emacs Lisp standards: it essentially replaces the old hook, advises and overrides-based activation mechanism by a cleanly delimited and defined minor mode.

The main visible change should be that if the old activation mechanism will keep working but generate a deprecation warning.  We advise that after updating the package, you modify your config to use the new interface:

``` elisp
;; This is deprecated:
;; (add-hook 'outline-minor-mode-hook 'outshine-hook-function)
;; 
;; Use this instead:
(add-hook 'outline-minor-mode-hook 'outshine-minor-mode)
```

We expect the transition to be flawless, but please report any issues or incompatibility to the [Outshine issue tracker on Github](https://github.com/alphapapa/outshine/issues).

**Comment by alphapapa** [comment]
Thanks, Thibault, that all sounds great.  I'm very happy that Outshine is proceeding well in your hands, and I can remain mostly hands-off for now.  :)  Let me know what you need me to do.

**Comment by yuhan0** [comment]
Just wondering, should we drop the `minor` qualifier and just call it `outshine-mode`? Hardly any other minor modes include it in their names, and there's no major mode to distinguish from unlike the case of `outline-minor-mode`.

**Comment by yuhan0** [comment]
Huh, just checked and it turns out quite a few minor modes do in fact name themselves that way. Still, it might be a good idea to visually distinguish a little from `outline-minor-mode`? Feel free to ignore 😅

**Comment by vv111y** [comment]
Hi all,
I want to do a spacemacs layer for this (outshine + outorg + navi), I'm happy to see this happen. This is a great package, I love using it

**Comment by thblt** [comment]
> Still, it might be a good idea to visually distinguish a little from outline-minor-mode? 

I looked through Elisp doc and it's neither a recommendation or an explicit convention for minor mode names to end in `-minor-mode`.  I've implemented those changes.

**Comment by thblt** [comment]
@tj64 (and others): what's the library you use to extract comments in outshine.el into a `README.md`?  I tried el2markdown, it doesn't really work with those comments.

**Comment by tj64** [comment]
@thblt:
I think I narrowed to that 1st level headline and transformed it to orgmode with outorg, then I exported the Org buffer to GitHub markdown, there is a very good org exporter for that (ox-gfm ???, not sure about the exact name).

Note that with prefix argument you can insert export headers when calling 'edit with outorg', see docstring of cmd

**Comment by thblt** [comment]
@tj64 Thanks, I ended up doing this:

```
(progn
  (find-file "outshine.el")
  (outorg-edit-as-org)
  (kill-ring-save (point-min) (point-max))
  (kill-buffer)
  (delete-file "README.org")
  (find-file "README.org")
  (yank)
  (save-buffer)
  (kill-buffer))
```

which is roughly your solution, without the export since Github now understands Org.

**Comment by thblt** [comment]
@tj64  Another question: what's the use of the *Metadata* section, which is only made of org properties? Can something read those?

**Comment by thblt** [comment]
I'm merging this. It should work, the remaining issues are cosmetic and can be handled separately.  I expect a 3.1 (or 3.0.1, whatever) really soon, but let's release the bulk of Outshine 3 first.  

Thanks all for your feedback and your help in this huge refactoring, I hope I'm not breaking your Emacs setups!

**Comment by tj64** [comment]
@thblt:
I once wrote a fairly long and detailed post on the Emacs Devel Mailinglist about modernizing some elisp conventions (using outshine), pointing to some pathologies (IMO).

One was the free text meta-data, which could be standardized and more machine readable by using outshine / org properties.

But this idea has not gained any attention, so you can transform it's content to free text respecting the elisp conventions, and delete that section if you want.

**Comment by tj64** [comment]
@all:
I'm impressed by this work and happy to see how the 'outshine suite' is improved and becoming even more useful.

Thank you for that big effort!

File: outshine.el [hunk 8]
```diff
@@ -523,11 +523,6 @@ them set by set, separated by a nil element.  See the example for
 (defvar outshine-agenda-old-org-agenda-files nil
   "Storage for old value of `org-agenda-files'")
 
-;;;; Hooks
```
Comment (alphapapa) [inline outshine.el]: This is replaced by the hook that's defined by `define-minor-mode`, right?

**Comment by alphapapa** [comment]
@tj64 It would never have happened without your originally writing the package!  Thank you, too!

**Comment by yuhan0** [comment]
@vv111y 
> I want to do a spacemacs layer for this (outshine + outorg + navi), I look forward to using this major update. This is a great package, I love using it

Hey I'm a Spacemacs user as well and wrote my own private outshine layer with lots of features, will most likely clean it up and submit to the official repo once things have settled down a little.

Thanks to everyone else too! Outshine really became an indispensible part of my config -  can't imagine going back to editing code without an outline structure now :)

**Comment by tj64** [comment]
@yuhan0
After a long (involuntary) abstinence from Emacs, I'm starting to use it again, and I switched to spacemacs, so I would really love to see an outshine spacemacs layer (including outorg/navi-mode?)!

**Comment by vv111y** [comment]
@yuhan0 
I looked at your fork and couldn't find your layer. Where is it? I would like to compare with what I have, thanks

**Comment by vv111y** [comment]
@yuhan0 
Ok I have never done a PR, let alone for spacemacs; I don't if I'm doing it right.  
I have my preliminary code in branch ```add-outshine-layer```, and I placed it in ```/layers/tools```

[outshine layer](https://github.com/vv111y/spacemacs/tree/add-outshine-layer/layers/%2Btools/outshine)  

I believe I have done the basics so maybe you would prefer to work off of this.  I have also included a transient state hydra in keybindings.el (for now). 
See what you think

**Pull request merged**